### ISSUE TITLE
Implement output operator '<<' for dual expressions

### ISSUE THREAD

**Issue by allanleal** [autodiff/autodiff#65]
We need to support output of `dual` expressions, for example:

~~~c++
dual a = 10.0;
std::cout << sin(a) << std::endl;
~~~

**Comment by ram-nad** [comment]
Sir, I would like go work on these. Any suggestions on output format?

**Comment by allanleal** [comment]
Hi @ram-nad ; great, thanks for your interest in helping this project. 

I think `operator<<` is defined for `Dual<T, G>`, where `T` and `G` are the template types for the value and gradient of the dual number. It is remaining to implement `operator<<` for expressions. I think we need something like the following:

~~~c++
template<typename R, enableif<isExpr<R>>...>
std::ostream& operator<<(std::ostream& out, R&& expr)
{
    out << val(expr);
}
~~~

Now I can't remember if `isExpr<R>` is `true` for `Dual` as well. If so, we only need this one and remove the `operator<<` overload for `Dual`.

**Comment by supersega** [comment]
@allanleal yes, it’s true

**Comment by ram-nad** [comment]
Sir, can we display a bit information about a operation applied? Or the value of expression is sufficient?
Also sir, I was studying the code and could not understand the use of the function:

```cpp
constexpr void assign(Dual<T, G>& self, U&& other, Dual<T, G>& tmp)
```
Can you please help me?

**Comment by ram-nad** [comment]
Also sorry for being so late.

**Comment by allanleal** [comment]
The entire function is:

~~~c++

template<typename T, typename G, typename U>
constexpr void assign(Dual<T, G>& self, U&& other, Dual<T, G>& tmp)
{
    static_assert(isExpr<U> || isNumber<U>);

    // ASSIGN AN UNARY EXPRESSION: self = func(expr)
    if constexpr (isUnaryExpr<U>) {
        using Op = OperatorType<U>;
        assign(self, other.r, tmp);
        apply<Op>(self);
    }
    // ASSIGN AN ADDITION EXPRESSION: self = expr + expr
    else if constexpr (isAddExpr<U>) {
        assign(self, other.r, tmp);
        assignAdd(self, other.l, tmp);
    }
    // ASSIGN A MULTIPLICATION EXPRESSION: self = expr * expr
    else if constexpr (isMulExpr<U>) {
        assign(self, other.r, tmp);
        assignMul(self, other.l, tmp);
    }
    // ASSIGN A POWER EXPRESSION: self = pow(expr, expr)
    else if constexpr (isPowExpr<U>) {
        assign(self, other.l, tmp);
        assignPow(self, other.r, tmp);
    }
    // ALL OTHER EXPRESSIONS
    else {
        assign(tmp, other);
        assign(self, tmp);
    }
}
~~~

The confusion might be because of the other similar function:

~~~c++
template<typename T, typename G, typename U>
constexpr void assign(Dual<T, G>& self, U&& other)
~~~

The reason for these two versions is that if during the expression evaluation a temporary is created, we can then make use of this for the rest of the expression evaluation. That's why we have the argument `Dual<T, G>& tmp` in the previous one. It's all recursive and it can be (naturally) a bit difficult to understand at first. 

Let me know if this explanation is not enough (or not what you are asking about).

As to the output, we should output it like if the number was a `double` value. We don't output the expression. We could have a print function for the expression, but what I usually do is to force a compilation error so that the expression is output in the compilation log. Something like this:

~~~c++
dual x, y, z;
bool b = x*y*exp(z); # compile error because the expression cannot be converted to bool
~~~

**Comment by ram-nad** [comment]
Sir, thanks for the explanation. Sir, while printing anyway we will need to evaluate the whole expression. So, can we store the value in the Dual or expression itself so it can work like cache, if the value is used immediately.

**Comment by supersega** [comment]
I'm sorry to get in.
> So, can we store the value in the Dual or expression itself so it can work like cache, if the value is used immediately.

But what the point?

**Comment by allanleal** [comment]
I don't think we need the cache. It can complicate things unnecessarily. Also, these expressions are fast evaluated. Differently from reverse mode, which can generate a huge expression tree (at runtime), forward mode produces several smaller ones (at compile time) that get evaluated very fast.

**Comment by ram-nad** [comment]
Okk sir.

**Comment by ludkinm** [comment]
> Hi @ram-nad ; great, thanks for your interest in helping this project.
> 
> I think `operator<<` is defined for `Dual<T, G>`, where `T` and `G` are the template types for the value and gradient of the dual number. It is remaining to implement `operator<<` for expressions. I think we need something like the following:
> 
> ```c++
> template<typename R, enableif<isExpr<R>>...>
> std::ostream& operator<<(std::ostream& out, R&& expr)
> {
>     out << val(expr);
> }
> ```
> 
> Now I can't remember if `isExpr<R>` is `true` for `Dual` as well. If so, we only need this one and remove the `operator<<` overload for `Dual`.

With a return for the ``std::ostream&``
```c++
template<typename R, enableif<isExpr<R>>...>
std::ostream& operator<<(std::ostream& out, R&& expr)
{
    return out << val(expr);
}
```
